src/rdkafka_topic.c

```diff
@@ -552,15 +570,57 @@ static int rd_kafka_toppar_leader_update2 (rd_kafka_itopic_t *rkt,
         rktp = rd_kafka_toppar_s2i(s_rktp);
 
         rd_kafka_toppar_lock(rktp);
-        r = rd_kafka_toppar_leader_update(rktp, leader_id, rkb);
+
+        if (leader != NULL &&
+            rktp->rktp_broker != leader &&
+            rktp->rktp_leader_id == leader_id) {
+                rd_kafka_dbg(rktp->rktp_rkt->rkt_rk, TOPIC, "BROKER",
+                        "Topic %s [%"PRId32"]: leader %"PRId32" unchanged, "
+                        "not migrating away from preferred replica %"PRId32,
+                        rktp->rktp_rkt->rkt_topic->str,
+                        rktp->rktp_partition,
+                        leader_id, rktp->rktp_broker_id);
+                r = 0;
+        } else
+                r = rd_kafka_toppar_broker_update(rktp, leader_id, 
+                                                  leader, rd_true);
+
         rd_kafka_toppar_unlock(rktp);
 
-	rd_kafka_toppar_destroy(s_rktp); /* from get() */
+        rd_kafka_toppar_destroy(s_rktp);  /* from get() */
 
 	return r;
 }
 
 
+/**
+ * @brief Revert the topic+partition delegation to the leader.
+ * @returns 1 if the broker delegation was changed, -1 if the broker
+ *          delegation was changed and is now undelegated, else 0.
+ *
+ * @locks caller must have rd_kafka_toppar_lock(rktp)
+ *        and NOT rd_kafka_topic_wrlock(rkt)
+ * @locality any
+ */
+int rd_kafka_toppar_delegate_to_leader (rd_kafka_toppar_t *rktp) {
+        rd_kafka_broker_t *leader;
+
+        rd_assert(rktp->rktp_leader_id != rktp->rktp_broker_id);
+
+        rd_kafka_dbg(rktp->rktp_rkt->rkt_rk, TOPIC, "BROKER",
+                "Topic %s [%"PRId32"]: Reverting from preferred "
+                "replica %"PRId32" to leader %"PRId32,
+                rktp->rktp_rkt->rkt_topic->str, rktp->rktp_partition,
+                rktp->rktp_broker_id, rktp->rktp_leader_id);
+
+        leader = rd_kafka_broker_find_by_nodeid(rktp->rktp_rkt->rkt_rk,
```

Comment:
this reference, if not NULL, is never released.
Call rd_kafka_broker_destroy after broker_update

Reply:
ouch, that's a bad mistake.